kernel/aster-nix/src/syscall/sched_getaffinity.rs

```diff
@@ -0,0 +1,55 @@
+// SPDX-License-Identifier: MPL-2.0
+
+use core::mem;
+
+use super::SyscallReturn;
+use crate::{
+    prelude::*,
+    process::{process_table, Pid},
+    util::write_val_to_user,
+};
+
+pub fn sys_sched_getaffinity(
+    pid: Pid,
+    cpuset_size: usize,
+    user_mask_ptr: Vaddr,
+) -> Result<SyscallReturn> {
+    // Ensure cpuset_size is sufficient to store a cpu_set_t.
+    if cpuset_size < core::mem::size_of::<cpu_set_t>() {
+        return Err(Error::with_message(Errno::EINVAL, "invalid cpuset size"));
+    }
+
+    // Use match statement for clearer control flow especially when pid == 0 logic is implemented.
+    match pid {
+        0 => {
+            // TODO: Get the current thread's CPU affinity
+            // Placeholder for future implementation.
+        }
+        _ => {
+            // Check if the process exists, in a Rust idiomatic way using match for clarity.
+            match process_table::get_process(&pid) {
+                Some(_process) => { /* Placeholder if process-specific logic needed */ }
+                None => return Err(Error::with_message(Errno::ESRCH, "process does not exist")),
+            }
+        }
+    }
+
+    // Assuming all CPUs are available; construct the cpu_set_t accordingly.
+    let result = cpu_set_t {
+        __bits: [usize::MAX; CPU_SETSIZE / __NCPUBITS],
```

Comment:
I think the number of bit 1 should be equal to the number of CPUs when all CPUs are available to schedule.

Reply:
Thanks for the advice! I have revised this part.

The function `fn get_num_cpus() -> usize` provides a mechanism to determine the actual number of CPUs available on a system at runtime.

The `cpu_set_t::new` function creates a precise CPU affinity mask based on num_cpus, which signifies the desired number of available CPUs for scheduling.
This implementation iterates over CPU indices up to `num_cpus` or `CPU_SETSIZE`. For each index, it sets a corresponding bit to 1 in the `__bits` array, reflecting the actual CPU availability.